clang/test/SemaSYCL/sycl-restrict.cpp

```diff
@@ -133,12 +152,76 @@ void usage(myFuncDef functionPtr) {
     b.f(); // expected-error {{SYCL kernel cannot call a virtual function}}
 
   Check_RTTI_Restriction::kernel1<class kernel_name>([]() { // expected-note 3{{called by 'usage'}}
-  Check_RTTI_Restriction::A *a;
-  Check_RTTI_Restriction::isa_B(a); }); // expected-note 6{{called by 'operator()'}}
+    Check_RTTI_Restriction::A *a;
+    Check_RTTI_Restriction::isa_B(a); // expected-note 6{{called by 'operator()'}}
+  });
 
-  __float128 A; // expected-error {{__float128 is not supported on this target}}
+  // ======= Float128 Not Allowed in Kernel ==========
```

Comment:
Does it diagnose if some struct has a field with unsupported type and this variable with this struct type is declared inside device code?

Reply:
I will check it and see.
Right now, the only SYCL code the type checks struct members is the recursive type check code that is  in CheckSYCLType over in SemaSYCL.cpp, the one called by the AST visitor.

Reply:
Hmm - so struct members are problematic.   They are overlooked.  

I can recursively check them, that works well enough.  It correctly issues an error if the violating struct is used in a kernel context, when that error is issued it is on the struct declaration itself. Great.   But  it's  confusing, because _declaring_ the struct isn't really a problem, _using_ it in a kernel context is. 

Right now, the checks we have in the AST Visitor, since they don't use the deferred diagnostics, correctly recurse struc members and  output an error on its declaration AND they also issue the `diag::note_sycl_used_here`  to guide the user back to the usage that triggers it.  But this same approach isn't practical with deferred diagnostics. 

Let me investigate some more.  I can think of a handful of approaches.

Reply:
OK - I brought some of the recursive struct code out of CheckSYCLType. Tidied it up and got the notes to emit as a follow on deferred diagnostic. Struct member are working now.